### ISSUE TITLE
Python2/urlparse and str/unicode

### ISSUE THREAD

**Issue by bsolomon1124** [python/typeshed#3844]
I'm wondering if it may make sense to use `Union[str, unicode]` in place of `str` in several places in `stdlib/2/urlparse.pyi`.  Specifically, the members of `_ParseResult` and `_SplitResult` may make sense as basestring.  The urlparse docs do show examples as `str` (unless we assume `unicode_literals` has been imported behind the scenes 😉 ), but they also make notes such as "If the URL is decomposed before parsing, or is not a Unicode string, no error will be raised."

And more directly, `urlunparse` is more than happy to take a namedtuple of unicode and give back a unicode URL:

```python
>>> from __future__ import unicode_literals
>>> from urlparse import ParseResult, urlunparse
>>> ParseResult(scheme='http', netloc='www.google.com', path='/me/', params='', query='foo=bar', fragment='')
ParseResult(scheme=u'http', netloc=u'www.google.com', path=u'/me/', params=u'', query=u'foo=bar', fragment=u'')
>>> urlunparse(ParseResult(scheme='http', netloc='www.google.com', path='/me/', params='', query='foo=bar', fragment=''))
u'http://www.google.com/me/?foo=bar'
```

**Comment by srittau** [comment]
Members that can be either should indeed be typed as an union of the two. For argument types, `Union[str, unicode]` should be expressed as `Text`.

**Comment by bsolomon1124** [comment]
Same thing for [`kwargs` of `threading.Thread`](https://github.com/python/typeshed/blob/master/stdlib/2and3/threading.pyi), it looks like.  Will open a PR as soon as I have time.

**Comment by lladhibhutall** [comment]
@bsolomon1124 Are you working on this? I can take a stab at it and submit a patch

**Comment by bsolomon1124** [comment]
Please feel free to @lladhibhutall , or I should be able to get to it this weekend.

**Comment by pmhahn** [comment]
I think ddb47de broke `threading.Thread` with Python 2:
```python
from typing import Any, Dict, Mapping, Text

def wrap(**kwargs):  # type: (**Any) -> None
  nested(kwargs)  # `kwargs` is `Dict[str, Any]`

def nested(kwargs):  # type: (Mapping[Text, Any]) -> None
  pass

wrap(arg='')
```
leads to this error now as `Text` is just an alias for `unicode` with Python 2:
```
error: Argument 1 to "nested" has incompatible type "Dict[str, Any]"; expected "Mapping[unicode, Any]"
```
Changing it back from `Text` to `str` fixes that case and also seems to accept `unicode`:
```
wrap(**{u"\u1234": ""})
```

**Comment by srittau** [comment]
@pmhahn I am unsure how your example relates to the commit. Do you mean we need to revert `Mapping[Text, Any]` in `Thread.__init__()` back to ` Mapping[str, Any]`? This makes sense to me, since the former is a sub-type of the latter, but not vice versa.

**Comment by pmhahn** [comment]
@srittau Yes, from my current understanding the change to `Thead.__init__()` should be reverted.  ddb47de changed the code this way:
```
def __init__(self

**Comment by pmhahn** [comment]
Yes, from my current understanding  ddb47de should be reverted - at least the part for `threading.Thread`, as it contains the following change:
```diff
--- a/stdlib/2and3/threading.pyi
+++ b/stdlib/2and3/threading.pyi
@@ -67 +68 @@ class Thread:
-                     kwargs: Mapping[str, Any] = ...) -> None: ...
+                     kwargs: Mapping[Text, Any] = ...) -> None: ...
```
This in the "Python 2 only" part, so the function now requires a `Dict[unicode, Any]` and no longer accepts `Dict[str, Any]`. The later is created implicitly if you use a function with `def f(**kwargs)`, which is a pain.

I'm still unsure how to express an API in Python 2, which both accepts `bytes` and `unicode`:
1. is `str` enough?
2. or `Union[unicode, str]`?
3. or `Text`?

From my previous experiments it should be 1., but I'm still not 100% sure and did not find an explicit documentation.yet.

**Comment by srittau** [comment]
`Text` is generally the right type in such cases. `Text` in arguments is an alias for `Union[str, unicode]`. (Outside of arguments, its interpretation as either `Union[str, unicode]` or just `unicode` is left to the type checker.)

**Comment by pmhahn** [comment]
Sorry, I still don't get it: Quoting from above:

> `Mapping[Text, Any]` is a sub-type of ` Mapping[str, Any]`, but not vice versa.

but later

> `Text` in arguments is an alias for `Union[str, unicode]`. (Outside of arguments, its interpretation as either `Union[str, unicode]` or just `unicode` is left to the type checker.)

I think I still have to re-read [Variance of generic types](https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types) many more times until I get it right. Because of that lack of understanding I'm unsure if ddb47de should be reverted completely or only partially.

**Comment by JelleZijlstra** [comment]
Mapping is invariant in its key type, so neither type is a subtype of the other.

**Comment by hauntsaninja** [comment]
What is Python 2?
